src/ray/raylet/node_manager.cc

```diff
@@ -1292,33 +1292,20 @@ void NodeManager::ProcessRegisterClientRequestMessage(
     rpc::JobConfig job_config;
     job_config.ParseFromString(message->serialized_job_config()->str());
 
-    // Send the reply callback only after registration fully completes at the GCS.
-    auto cb = [this,
-               worker_ip_address,
-               worker_id,
-               pid,
-               job_id,
-               job_config,
-               entrypoint,
-               send_reply_callback = std::move(send_reply_callback)](const Status &status,
-                                                                     int assigned_port) {
-      if (status.ok()) {
-        rpc::Address driver_address;
-        // Assume raylet ID is the same as the node ID.
-        driver_address.set_raylet_id(self_node_id_.Binary());
-        driver_address.set_ip_address(worker_ip_address);
-        driver_address.set_port(assigned_port);
-        driver_address.set_worker_id(worker_id.Binary());
-        auto job_data_ptr = gcs::CreateJobTableData(
-            job_id, /*is_dead*/ false, driver_address, pid, entrypoint, job_config);
-
-        RAY_CHECK_OK(gcs_client_->Jobs().AsyncAdd(
-            job_data_ptr,
-            [send_reply_callback = std::move(send_reply_callback), assigned_port](
-                Status status) { send_reply_callback(status, assigned_port); }));
-      }
-    };
-    RAY_UNUSED(worker_pool_.RegisterDriver(worker, job_config, std::move(cb)));
+    // Prepare the JobTableData for the GCS. However we will not send it to GCS yet,
+    // because we need to wait for the worker to announce its port first. Sending in
+    // `ProcessAnnounceWorkerPortMessage`.
+    rpc::Address driver_address;
+    // Assume raylet ID is the same as the node ID.
+    driver_address.set_raylet_id(self_node_id_.Binary());
+    driver_address.set_ip_address(worker_ip_address);
+    // driver_address.port is not set! It will be set in ProcessAnnounceWorkerPortMessage
+    driver_address.set_worker_id(worker_id.Binary());
+    auto job_data_ptr = gcs::CreateJobTableData(
+        job_id, /*is_dead=*/false, driver_address, pid, entrypoint, job_config);
+    worker->SetDriverJobTableData(job_data_ptr);
+
+    RAY_UNUSED(worker_pool_.RegisterDriver(worker, job_config, send_reply_callback));
```

Comment:
I think we can just construct JobTableData inside AnnouncePort, whatever information you need, you can include in the AnnouncePortRequest (e.g. worker_ip_address).

Reply:
The problem is `entrypoint` which is passed from core_worker to raylet in `RegisterClientRequest`. It's not making the most sense to pass in `AnnouncePortRequest`.

Reply:
I guess this is just because the current rpc is named `AnnouncePort`? If it's named something like `AnnounceReady`, it makes sense?

I just felt that we now store something permanently for something that's only needed locally.

Reply:
It's not only the problem of naming. the field `entrypoint` is passed in `RegisterClientRequest` so we need to keep it until we got the AnnouncePort. I can free the memory after it's used in AnnouncePort. Does that sound good?

Reply:
OK now I understand you want to put the `entrypoint` in AnnouncePortRequest. Ok will do that. Was misreading your comments lol